### DISCUSSION TITLE
bottled vs. built from source openblas

### DISCUSSION THREAD

**Discussion by zlscherr** [Homebrew/discussions#54] [Everyday usage]
Hi All,

   I had a quick comment to the community regarding the bottled openblas library that I was hoping somebody could weight in on.  If, on my 2018 MBP, I run:

```
brew install openblas
brew install numpy
python3 -c "import numpy as np; x=np.random.randn(1000000,2); y=np.random.randn(2,2); x.dot(y)"
```

then my program crashes with a segmentation fault.  However, if instead I do

```
brew install openblas --build-from-source
brew install numpy
python3 -c "import numpy as np; x=np.random.randn(1000000,2); y=np.random.randn(2,2); x.dot(y)"
```

then the program runs without any problems.

My guess is that the bottled version was built with a different TARGET architecture than what my laptop has.  Does anybody know if there is any way around having to build from source every time? It might be worthwhile to note this in the info for openblas that the bottled version may not work on all TARGETS.

**Comment by zlscherr** [comment]
I learned from one of the maintainers of the Conda-forge openblas that the problem is because openblas is compiled with DYNAMIC_ARCH=1 but no TARGET= .  So it seems that this is an easy fix just by adding a TARGET telling openblas the oldest CPU that it should be compiled against, and then it will change the kernel at runtime based on what the current architecture is.

**Reply by SMillerDev** [reply]
Could you make a pull request for this?

**Reply by zlscherr** [reply]
Working on that now.  I'm new at this, but I will do the pull request shortly.  My apologies if I do anything incorrectly.

**Reply by SMillerDev** [reply]
I'd much rather help fix a PR then having to redo someone else's research from a forum post. Tag me when you're done and I'll help you get it merged.

**Reply by zlscherr** [reply]
Thank you, I just did submitted the pull request.

**Comment by zlscherr** [comment]
I mentioned this over at the PR, but this is an openBLAS issue: https://github.com/xianyi/OpenBLAS/issues/2728 that they are aware of and should be fixed in 0.3.11.

I still think it's worth setting TARGET because leaving it out could still cause problems on older architectures if the bottled libraries were built on newer ones.
